### PR TITLE
Audit logging for Dicom update

### PR DESCRIPTION
**Pull request by arunmk-ms**
## Description
Enable audit logging for Dicom update feature.

## Related issues
Addresses [US 102012](https://microsofthealth.visualstudio.com/Health/_workitems/edit/102012).

## Testing
Verified manually in local as well as in Dgrep audit tables.

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions.App/host.json [hunk 1]
```diff
@@ -100,5 +100,8 @@
         "MaxTimeInterval": "00:00:20"
       }
     }
+  },
+  "DicomService": {
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions.App/host.json:104]: This should definitely be in the `local.settings.json`

File: src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs [hunk 2]
```diff
@@ -116,7 +117,7 @@ public interface IDicomOperationsClient
     /// <paramref name="updateSpecification"/> is <see langword="null"/>.
     /// </exception>
     /// <exception cref="OperationCanceledException">The <paramref name="cancellationToken"/> was canceled.</exception>
-    Task<OperationReference> StartUpdateOperationAsync(Guid operationId, UpdateSpecification updateSpecification, int partitionKey, CancellationToken cancellationToken = default);
+    Task<OperationReference> StartUpdateOperationAsync(Guid operationId, UpdateSpecification updateSpecification, int partitionKey, string endpoint, CancellationToken cancellationToken = default);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs:120]: Why is this part of the function input? Can this not be part of the function app's config?

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 3]
```diff
@@ -41,6 +45,21 @@ public partial class UpdateDurableFunction
         logger = context.CreateReplaySafeLogger(EnsureArg.IsNotNull(logger, nameof(logger)));
 
         UpdateCheckpoint input = context.GetInput<UpdateCheckpoint>();
+        Uri uri = new Uri(input.Endpoint, UriKind.Absolute);
+        var callerClaims = new List<KeyValuePair<string, string>>
+        {
+            new ("operation_id", context.GetOperationId().ToString()),
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:51]: I think you can just use `InstanceId` unless you want a specific `Guid` format

Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:51]: Should this key be defined somewhere?

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 4]
```diff
@@ -41,6 +45,21 @@ public partial class UpdateDurableFunction
...
+        _auditLogger.LogAudit(
+            AuditAction.Executing,
+            AuditEventSubType.UpdateStudy,
+            uri,
+            null,
+            Activity.Current?.RootId,
+            null,
+            callerClaims,
+            null);
+        HttpStatusCode auditStatusCode = HttpStatusCode.BadRequest;
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:62]: Both `HttpStatusCode` and `Url` are in this log? Is this an appropriate log for this kind of operation?

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Dicom.Functions.App/host.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions.App/host.json]: Removed the same. I had initially added it for using in local environment.

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 5]
```diff
@@ -42,6 +46,21 @@ public partial class UpdateDurableFunction
...
+        };
+        _auditLogger.LogAudit(
+            AuditAction.Executing,
+            AuditEventSubType.UpdateStudy,
+            _dicomServiceOptions.Endpoint,
+            null,
+            Activity.Current?.RootId,
+            null,
+            callerClaims,
+            null);
```
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:61]: `statusCode` should be null when auditAction is `Executing`. `callerIpAddress` and `customHeaders` are the other two fields that are optional.

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 6]
```diff
@@ -124,6 +143,16 @@ public partial class UpdateDurableFunction
             {
                 logger.LogInformation("Update operation completed successfully");
             }
+            auditStatusCode = HttpStatusCode.OK;
         }
+        _auditLogger.LogAudit(
+            AuditAction.Executed,
+            AuditEventSubType.UpdateStudy,
```
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs]: Good call out. I have changed it to `update-study-task`

File: src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs (hunk 2 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs:120]: We have the endpoint available in web, but not in functions. So the endpoint is grabbed from the requestContext when the user queues an update operation

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs (hunk 3 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs]: We are not using this key anywhere else. So it is safe to hardcode for now.

File: src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs (hunk 2 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Operations/IDicomOperationsClient.cs:120]: Oh, so we need to log the entire URL? Not just the host?

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 7]
```diff
@@ -41,6 +45,21 @@ public partial class UpdateDurableFunction
...
 
         UpdateCheckpoint input = context.GetInput<UpdateCheckpoint>();
+        Uri uri = new Uri(input.Endpoint, UriKind.Absolute);
+        var callerClaims = new List<KeyValuePair<string, string>>
+        {
+            new ("instance_id", context.InstanceId),
+        };
+        _auditLogger.LogAudit(
+            AuditAction.Executing,
+            AuditEventSubType.UpdateStudyTask,
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:55]: We can call UpdateStudyOperation

**Pull request closed without merging**